### PR TITLE
Support sparse gradients in DistributedDataParallel

### PR DESCRIPTION
**Pull request by pietern**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#19443 Support sparse gradients in DistributedDataParallel**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15007365/)
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; #19146 Add sparse tensor allreduce&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D14889547/)

This adds support for sparse gradients to the reducer as well as to
the DistributedDataParallel wrapper. Note that an out of band signal
is needed whether or not a dense parameter (e.g. an embedding) is
expected to receive a sparse gradient or not. This information is
passed to the bucket assignment computation routine and the reducer as
a vector of booleans. Every parameter for which we expect a sparse
gradient is assigned its own bucket, as we cannot easily group
multiple unrelated sparse tensors.

Differential Revision: [D15007365](https://our.internmc.facebook.com/intern/diff/D15007365/)

### CODE REVIEW

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 1]
```diff
@@ -495,6 +600,14 @@ std::vector<std::vector<size_t>> compute_bucket_assignment_by_size(
   for (size_t i = 0; i < tensors.size(); i++) {
     const auto& tensor = tensors[i];
     AT_ASSERTM(!tensor.is_sparse(), "No support for sparse tensors.");
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.cpp]: Is this check still necessary? Or will it be possible that `is_sparse()` returns different values for tensor and its gradient?

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 2]
```diff
@@ -61,6 +64,11 @@ Reducer::Reducer(
       AT_ASSERTM(
           replicas_[replica_index].size() == replicas_[0].size(),
           "Model replicas must have an equal number of parameters.");
+      AT_ASSERTM(
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.cpp]: Q: (not necessary to make any changes here) as `expect_sparse_gradients_ ` should be the same for all replicas, do we still need a 2D list there? (same for `bucket_indices`)

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 3]
```diff
@@ -137,6 +150,64 @@ Reducer::Reducer(
...
+    size_t variable_index) {
+  const auto& bucket_index = variable_locators_[variable_index];
+  auto& bucket = buckets_[bucket_index.bucket_index];
+  auto& replica = bucket.replicas[replica_index];
+  auto& variable = replica.variables[bucket_index.intra_bucket_index];
+  auto& grad = variable.grad();
+  AT_ASSERTM(
+      grad.options().layout() == c10::kSparse,
+      "Expected variable to have sparse gradient.");
+  AT_ASSERTM(grad.defined(), "Expected sparse gradient to be defined.");
```
Comment (mrshenli) [inline torch/csrc/distributed/c10d/reducer.cpp]: shall we do this for dense as well? Or is there any reason for the inconsistency between dense and sparse?

**Review by mrshenli — COMMENTED** [review]
Shall we add some tests?

File: torch/nn/parallel/distributed.py [hunk 4]
```diff
@@ -293,24 +293,51 @@ def _ddp_init_helper(self):
...
+            ] for replica in self._module_copies]
+
+        # Build list of parameters.
+        parameters = [
+            list(parameter for _, parameter in replica)
+            for replica in modules_and_parameters]
+
+        # Checks if a module will produce a sparse gradient.
+        def produces_sparse_gradient(module):
+            if isinstance(module, torch.nn.Embedding):
```
[the remark this answers is not in the document]
Comment (pietern) [reply torch/nn/parallel/distributed.py]: The embedding itself is dense so we have to differentiate between the parameters that produce dense gradients (the default) and the ones that produce sparse gradients. We can start with just these and deal with user specified layers that produce sparse gradients later.

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 1 above)
Comment (pietern) [reply torch/csrc/distributed/c10d/reducer.cpp]: Yes, the parameter itself will be dense.

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 2 above)
Comment (pietern) [reply torch/csrc/distributed/c10d/reducer.cpp]: It should be the same, indeed. I rather have the additional check here than deal with runtime errors where they are not the same. At least with this in place we can make those checks and it will barf if there is any asymmetry between replicas.

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 3 above)
Comment (pietern) [reply torch/csrc/distributed/c10d/reducer.cpp]: Good point. Added the layout assert.

**Comment by pietern** [comment]
@mrshenli Added a test case that confirms numerical equivalence between unwrapped and wrapped module.

**Review by mrshenli — APPROVED** [review]

**Comment by ezyang** [comment]
This diff stack broke ~all tests on master. Sample failure: https://circleci.com/gh/pytorch/pytorch/2038576?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console

**Comment by sheenG** [comment]
hello, I wonder if the modification is effective in these .whl files ? [torch_download](https://download.pytorch.org/whl/torch/)

**Comment by sheenG** [comment]
I got torch by `pip install torch==1.9.0` and I cannot set sparse=True in nn.Embedding under ddp.

**Comment by YeDeming** [comment]
I cannot set sparse=True in nn.Embedding under ddp on 1.9.1+cu102, too.
